### ISSUE TITLE
project plans

### ISSUE THREAD

**Issue by victron** [bolshevik/goke-GK7102-customizer#2]
Hi @bolshevik 
thank you for your job, I have same interest. Especially in running mod from SD.
some questions to you:
1. in `custom_firmware/data/` quite a lot of files. Does they all needed for mod running? From my experiments only `config_info.ini, .wifi_conn_info, mac.ini, myuid.bin` needs to generate `.config1`.
2. Are you going add sftp support? I could prepare pr.
3. any future plans? Did you integrated you camera with private cloud?

**Comment by bolshevik** [comment]
Hello @victron , 

1. These files were extracted from the modified firmware obtained from 4PDA. If we could confirm we need less, we could also clean up those not needed.

2. SFTP-like functionality is already covered with SSH+scp from my point of view. SFTP would need another program for the daemon to be running. If you would like we could add it but I would suggest to have it optional to have more free RAM.

3. At this point one could use the official SAP HD app to connect to the cam on the LAN, it is also possible to access RTSP or web interface if ports are forwarded outside of the LAN (but consider privacy and security) or if there is a VPN access to the LAN.

The cloud topic is pretty complex, it looks like smth similar to https://github.com/cnping/TUTK is used. At least there are many symbols referencing IOTC platform (see P2PTunnelAPIs.h) but probably of newer version and some library called `ubia`.

Regarding plans the only thing I was thinking of was to add iptables binaries to be able to better secure the device on the network, but if I am not mistaken, kernel modules were missing for it.

**Comment by victron** [comment]
Hi,
thank you for replay, I probably will split this topic into new issues.
just one note related to iptables, my mind it's not even need to think about it.
Firewall normally and should be implemented on wifi router. Wifi routers especially designed to create secure perimeter for home devices. Also NAT on routers provides additional security.
Of cause all above true if you are not connecting your camera to free wifi )))).

**Comment by milankragujevic** [comment]
Hi, finally someone mention **ubia**!

I have ACME IP1202, that seems to use some modified version of this protocol. 

It talks to m1.ubia.cn, m2.ubia.cn, m3.ubia.cn and m4.ubia.cn, as well as camera.kkmove.com (sends data via HTTP but the website is broken) and baidu.cn.

The only open port is 23, I scanned multiple times with very broad settings, nothing is open other than 23. No known password I tried for telnet works. I don't have a serial adapter right now to access it like that, so I'm looking for a way to access the video stream with ffmpeg.

I will then block it's access to the internet with a MikroTik router. 

I can't find any info on this, and I don't know enough to do this myself. Please help me.

By the way, all camera apps that work are cn.ubia.SOMETHING, from Maxwell Wang on Play Store, I'm using GuardVR app right now.w

**Comment by milankragujevic** [comment]
@bolshevik Can you help me?

**Comment by bolshevik** [comment]
@milankragujevic In this case this repository would not help a lot I think. In order to get more information you would need to extract the firmware: either through a programmer or please check if your camera has kind of an update server, maybe you could obtain the firmware from it. Then when taking a look at the firmware you would have two chances: either have a hash of the password and try to crack it or having the rootfs or cam application that might run some script from the SD card to hook into. Like I noted in the reader, that different firmwares might have different scripts they try to run. 

PS: you can extract the firmware with ‘binwalk’ usually.

**Comment by milankragujevic** [comment]
@bolshevik I only need a way to access the video stream, I don't care about the firmware. I assume, since the camera is talking to m1.ubia.cn, m2.ubia.cn, m3.ubia.cn and m4.ubia.cn (IOTC-compatible servers), it should be compatible with other cameras that also use IOTC. 

However, I haven't found ANY info on ubia.cn, or on IOTC-compatible viewing in unofficial apps. I am so desperate I'm thinking of running Android in a virtual machine on a server with the app open and screen capturing the display then cropping out just the camera video stream and live streaming that back to the house, but it seems so wasteful.

**Comment by bolshevik** [comment]
I see, the protocol is not open as you might have noticed. I was not able to find any details either. The biggest obstacle to understand how it works is that the android apps are having a pre-built version embedded. 

If I remember it correctly, the purpose of the library is to establish communication on the LAN, but this also might be wrong.

**Comment by milankragujevic** [comment]
@bolshevik Yes, it does talk over LAN. Here is a packet capture via MikroTik between the camera (192.168.1.230) and phone (192.168.1.202) - https://milankragujevic.com/uploads/camera-data.pcapng

The precompiled library is also available with header files for a few platforms here: https://github.com/cnping/TUTK/tree/master/TUTK_IOTC_Platform_14W36/Lib

Maybe you could find out more.

**Comment by milankragujevic** [comment]
Hi @bolshevik Can you help me? I sent the packet capture of communcation: https://milankragujevic.com/uploads/camera-data.pcapng

camera is 192.168.1.230
phone is 192.168.1.202

Model is ACME IP1202, also known as Southern Telecom SVC660.

FCC filing here: https://fccid.io/2ABV4-SVC660

I have verified internals are identical.

**Comment by bolshevik** [comment]
Hello @milankragujevic, sorry for the delay, I do not think I could help you here.

**Comment by milankragujevic** [comment]
That is sad to hear. Okay.

**Comment by neagix** [comment]
@milankragujevic did you make any progress on this? I gave a look at apps like https://apkpure.com/sap-hd/com.sap.SAPHD and https://apkpure.com/wise-vr/com.gg.WISEVR and they have quite a few references to TUTK, UBIA, IOTC.

The former uses `com.ubia.MyCamera` instead of `com.tutk.IOTC.MyCamera` (although both are present inside the APK).

The library is `libIOTCAPIs_ubia.so`, it does not seem too complicated to study. There is another project here using it: https://github.com/gb-6k-house/PIXPO-HD

In the TUTK_IOTC_Platform_14W36 folder of the [TUTK repository you linked](https://github.com/cnping/TUTK) there are unstripped versions.

**Comment by milankragujevic** [comment]
@neagix No, I bought Hikvision cameras that have RTSP enabled and a web UI. I don't have the skills and maybe even if I did I most certainly don't have the time. It was cheaper buying a new camera. Sorry :worried:
